mono/metadata/cominterop.c

```diff
@@ -2660,47 +2695,65 @@ cominterop_ccw_get_ids_of_names_impl (MonoCCWInterface* ccwe, gpointer riid,
 	static MonoClass *ComDispIdAttribute = NULL;
 	ERROR_DECL (error);
 	MonoCustomAttrInfo *cinfo = NULL;
-	int i,ret = MONO_S_OK;
-	MonoMethod* method;
+	int i, j, methodcount, pos, ret = MONO_S_OK;
+	MonoMethod* method = NULL;
 	gchar* methodname;
-	MonoClass *klass = NULL;
+	MonoClass* iface = ccwe->iface;
 	MonoCCW* ccw = ccwe->ccw;
-	MonoObject* object = mono_gchandle_get_target_internal (ccw->gc_handle);
+	MonoObject* object;
 
 	/* Handle DispIdAttribute */
 	if (!ComDispIdAttribute)
 		ComDispIdAttribute = mono_class_load_from_name (mono_defaults.corlib, "System.Runtime.InteropServices", "DispIdAttribute");
 
-	g_assert (object);
-	klass = mono_object_class (object);
-
 	if (!mono_domain_get ())
 		 mono_thread_attach (mono_get_root_domain ());
 
+	if (iface == mono_class_get_idispatch_class ()) {
+		object = mono_gchandle_get_target_internal (ccw->gc_handle);
+		g_assert (object);
+		iface = mono_object_class (object);
+	}
+
+	mono_class_setup_methods (iface);
+	methodcount = mono_class_get_method_count (iface);
+
 	for (i=0; i < cNames; i++) {
 		methodname = mono_unicode_to_external (rgszNames[i]);
```

Comment:
Is this memory freed?

Reply:
Can we free it? It seems to be potentially used by COM interop code for the lifetime of the class.

Reply:
methodname is a local, don't see where it is stored for he lifetime of the class?  Looking at the comments for mono_unicode_to_external it says:

"Callers must free the returned string."